tests/devices/qubit/test_initialize_state.py

```diff
@@ -0,0 +1,65 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for create_initial_state in devices/qubit."""
+
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.devices.qubit import create_initial_state
+from pennylane.operation import StatePrep
+
+
+class TestInitializeState:
+    """Test the functions in initialize_state.py"""
+
+    # pylint:disable=unused-argument,too-few-public-methods
+    class DefaultPrep(StatePrep):
+        """A dummy class that assumes it was given a state vector."""
+
+        num_wires = qml.operation.AllWires
+
+        def state_vector(self, wire_order=None):
+            return self.parameters[0]
+
+    @pytest.mark.all_interfaces
+    @pytest.mark.parametrize("interface", ["numpy", "jax", "torch", "tensorflow"])
+    def test_create_initial_state_no_state_prep(self, interface):
+        """Tests that create_initial_state works without a state-prep operation."""
+        state = create_initial_state([0, 1], like=interface)
+        assert qml.math.allequal(state, [[1, 0], [0, 0]])
+        assert qml.math.get_interface(state) == interface
+
+    @pytest.mark.all_interfaces
+    @pytest.mark.parametrize("interface", ["numpy", "jax", "torch", "tensorflow"])
+    def test_create_initial_state_with_state_prep(self, interface):
+        """Tests that create_initial_state works with a state-prep operation."""
+        prep_op = self.DefaultPrep(qml.math.array([1 / 2] * 4, like=interface), wires=[0, 1])
+        state = create_initial_state([0, 1], prep_operation=prep_op)
+        assert qml.math.allequal(state, [1 / 2] * 4)
+        assert qml.math.get_interface(state) == interface
+
+    def test_create_initial_state_with_BasisState(self):
+        """Tests that create_initial_state works with a real state-prep operator."""
+        prep_op = qml.BasisState([0, 1, 0], wires=[0, 1, 2])
+        state = create_initial_state([0, 1, 2], prep_operation=prep_op)
+        assert state[0, 1, 0] == 1
+        state[0, 1, 0] = 0  # set to zero to make test below simple
+        assert qml.math.allequal(state, np.zeros((2, 2, 2)))
+
+    def test_create_initialize_state_prefers_op_interface_over_like(self):
+        """Tests that the like argument is ignored when a prep-op is provided."""
+        prep_op = self.DefaultPrep([0, 0, 0, 1], wires=[0, 1])
+        state = create_initial_state([0, 1], prep_operation=prep_op, like="torch")
+        assert qml.math.get_interface(state) == "numpy"
```

Comment:
We should actually prefer `like` over prep-op interface, especially if the prep op is `numpy`.


Take for consideration `QuantumScript([qml.RX(torch.tensor(1.0), 0)], [qml.expval(qml.PauliZ(0))], prep_op)`.  WE will need the returned state to be `torch` or we will get errors whe applying the rotation gate.